pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/OffloadIndexBlockImpl.java

```
@@ -0,0 +1,336 @@
+    private static final Logger log = LoggerFactory.getLogger(OffloadIndexBlockImpl.class);
+
+    private static final int indexMagicWord = 0xDE47DE47;
+
+    private LedgerMetadata segmentMetadata;
+    private TreeMap<Long, OffloadIndexEntryImpl> indexEntries;
+
+    private final Handle<OffloadIndexBlockImpl> recyclerHandle;
+
+    private static final Recycler<OffloadIndexBlockImpl> RECYCLER = new Recycler<OffloadIndexBlockImpl>() {
+        @Override
+        protected OffloadIndexBlockImpl newObject(Recycler.Handle<OffloadIndexBlockImpl> handle) {
+            return new OffloadIndexBlockImpl(handle);
+        }
+    };
+
+    private OffloadIndexBlockImpl(Handle<OffloadIndexBlockImpl> recyclerHandle) {
+        this.recyclerHandle = recyclerHandle;
+    }
+
+    public static OffloadIndexBlockImpl get(LedgerMetadata metadata, List<OffloadIndexEntryImpl> entries) {
+        OffloadIndexBlockImpl block = RECYCLER.get();
+        block.indexEntries = Maps.newTreeMap();
+        entries.forEach(entry -> block.indexEntries.putIfAbsent(entry.getEntryId(), entry));
+        checkState(entries.size() == block.indexEntries.size());
+        block.segmentMetadata = metadata;
+        return block;
+    }
+
+    public static OffloadIndexBlockImpl get(InputStream stream) throws IOException {
+        OffloadIndexBlockImpl block = RECYCLER.get();
+        block.indexEntries = Maps.newTreeMap();
+        block.fromStream(stream);
+        return block;
+    }
+
+    public void recycle() {
+        segmentMetadata = null;
+        indexEntries.clear();
+        indexEntries = null;
+        if (recyclerHandle != null) {
+            recyclerHandle.recycle(this);
+        }
+    }
+
+    @Override
+    public OffloadIndexEntry getIndexEntryForEntry(long messageEntryId) {
+        if(messageEntryId > segmentMetadata.getLastEntryId()) {
+            log.warn("Try to get entry: {}, which beyond lastEntryId {}, return null",
```

this should throw an invalid argument exception. If a user is asking for an entry past the end of the ledger, that's a programmer error.